src/target/riscv/riscv-013.c

```diff
@@ -686,10 +704,92 @@ static int dmi_write_exec(struct target *target, uint32_t address,
 	return dmi_op(target, NULL, NULL, DMI_OP_WRITE, address, value, true, ensure_success);
 }
 
+static int dm_op_timeout(struct target *target, uint32_t *data_in,
+		bool *dmi_busy_encountered, int op, uint32_t address,
+		uint32_t data_out, int timeout_sec, bool exec, bool ensure_success)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_op_timeout(target, data_in, dmi_busy_encountered, op, address + dm->base,
+		data_out, timeout_sec, exec, ensure_success);
+}
+
+static int dm_op(struct target *target, uint32_t *data_in,
+		bool *dmi_busy_encountered, int op, uint32_t address,
+		uint32_t data_out, bool exec, bool ensure_success)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_op(target, data_in, dmi_busy_encountered, op, address + dm->base,
+			data_out, exec, ensure_success);
+}
+
+static int dm_read(struct target *target, uint32_t *value, uint32_t address)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_read(target, value, address + dm->base);
+}
+
+static int dm_read_exec(struct target *target, uint32_t *value, uint32_t address)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_read_exec(target, value, address + dm->base);
+}
+
+static int dm_write(struct target *target, uint32_t address, uint32_t value)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_write(target, address + dm->base, value);
+}
+
+static int dm_write_exec(struct target *target, uint32_t address,
+		uint32_t value, bool ensure_success)
+{
+	dm013_info_t *dm = get_dm(target);
+	if (!dm)
+		return ERROR_FAIL;
+	return dmi_write_exec(target, address + dm->base, value, ensure_success);
+}
+
+static bool check_dbgbase_exists(struct target *target)
+{
+	uint32_t next_dm = 0;
+	uint32_t count = 0;
+
+	LOG_TARGET_DEBUG(target, "Searching for DM with DMI base address (dbgbase) = 0x%x", target->dbgbase);
+	while (1) {
+		uint32_t current_dm = next_dm;
+		if (current_dm == target->dbgbase)
+			return true;
+		if (dmi_read(target, &next_dm, DM_NEXTDM + current_dm) != ERROR_OK)
+			break;
+		LOG_TARGET_DEBUG(target, "dm @ 0x%x --> nextdm=0x%x", current_dm, next_dm);
+		/* Check if it's last one in the chain. */
+		if (next_dm == 0) {
+			LOG_TARGET_ERROR(target, "Reach end of DM chain");
```

Comment:
I believe you can change this error print to debug. There is already a good and easy-to-understand error message in the function that calls `check_dbgbase_exists()`. 

If someone needs to dig deeper and troubleshoot what happened, they will most likely enable the verbose log (`-d`) and will be able to easily find out what happened.

```
LOG_TARGET_DEBUG(target, "Reach end of DM chain");
```

Reply:
I disagree. Both of these should be LOG_TARGET_ERROR. These errors will be very rare, so there's no risk of clutter. When the error in the caller does get printed, the user will always want to know immediately which of the two cases were hit. If you only want a single error message (which is not important to me in this case), then the correct way is to remove the error message in the calling function and add more details to these ones here.

Reply:
I don't mind both of those being LOG_TARGET_ERROR. It is a detail. 

(Still, the message about reaching the end of the DM chain feels redundant: If the search failed, it implicitly means we went through the list and reached the end without finding the item. On the other hand, if it is the very rare case hitting of maximum number of DMs on the DMI bus, then it justifies the extra error print about that unusual situation.)

Reply:
Let's just use LOG_TARGET_ERROR.
And to make it slightly more useful, I add count of DMs

https://github.com/riscv/riscv-openocd/compare/5cc1a4307e7e7981371530dcaf9c8aecef0192aa..b98698ac7c40125a72de14f5c67329a19c4b92d1

Reply:
Sounds good :)